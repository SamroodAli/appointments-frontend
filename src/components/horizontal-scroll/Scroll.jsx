import { useState } from 'react';
import { ScrollMenu } from 'react-horizontal-scrolling-menu';
import PropTypes from 'prop-types';

import { LeftArrow, RightArrow } from './arrow';
import Card from './card';

import useDrag from './useDrag';

const onWheel = (apiObj, ev) => {
  const isThouchpad = Math.abs(ev.deltaX) !== 0 || Math.abs(ev.deltaY) < 15;

  if (isThouchpad) {
    ev.stopPropagation();
    return;
  }

  if (ev.deltaY < 0) {
    apiObj.scrollNext();
  } else if (ev.deltaY > 0) {
    apiObj.scrollPrev();
  }
};

const Scroll = ({ items }) => {
  const {
    dragStart, dragStop, dragMove, dragging,
  } = useDrag();
  const handleDrag = ({ scrollContainer }) => (ev) => dragMove(ev, (posDiff) => {
    if (scrollContainer.current) {
      scrollContainer.current.scrollLeft += posDiff; //eslint-disable-line
    }
  });

  const [selected, setSelected] = useState('');
  const handleItemClick = (itemId) => () => {
    if (dragging) {
      return false;
    }
    return setSelected(selected !== itemId ? itemId : '');
  };

  return (
    <>
      <div className="example" style={{ paddingTop: '100px' }}>
        <div onMouseLeave={dragStop}>
          <ScrollMenu
            LeftArrow={LeftArrow}
            RightArrow={RightArrow}
            onWheel={onWheel}
            onMouseDown={() => dragStart}
            onMouseUp={() => dragStop}
            onMouseMove={handleDrag}
          >
            {items.map(({ id, name }) => (
              <Card
                name={name}
                itemId={id} // NOTE: itemId is required for track items
                key={id}
                onClick={handleItemClick(id)}
                selected={id === selected}
              />
            ))}
          </ScrollMenu>
        </div>
      </div>
    </>
  );
};

Scroll.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
  })).isRequired,
};

export default Scroll;
