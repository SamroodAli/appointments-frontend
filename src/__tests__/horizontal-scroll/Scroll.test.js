import renderer from 'react-test-renderer';
// import { render } from '@testing-library/react';
import Scroll from '../../components/horizontal-scroll/Scroll';
import Wrapper from '../../testWrapper';

beforeEach(() => {
  // IntersectionObserver isn't available in test environment
  const mockIntersectionObserver = jest.fn();
  mockIntersectionObserver.mockReturnValue({
    observe: () => null,
    unobserve: () => null,
    disconnect: () => null,
  });
  window.IntersectionObserver = mockIntersectionObserver;
});

test('Scroll snapshot testing', () => {
  const tree = renderer.create(
    <Wrapper>
      <Scroll onItemClick={jest.fn()} items={[]} />
    </Wrapper>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
