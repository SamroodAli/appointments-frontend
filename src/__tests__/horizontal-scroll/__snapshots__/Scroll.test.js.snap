// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroll snapshot testing 1`] = `
<div
  className="horizontal-scroll-container"
  onMouseLeave={[Function]}
>
  <div
    className="react-horizontal-scrolling-menu--wrapper "
    onMouseDown={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onWheel={[Function]}
  >
    <div
      className="arrow-left-container arrow-container"
    >
      <button
        className="arrow arrow-left"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        ◁
      </button>
    </div>
    <div
      className="react-horizontal-scrolling-menu--scroll-container scroll-container"
      onScroll={[Function]}
    />
    <div
      className="arrow-right-container arrow-container"
    >
      <button
        className="arrow arrow-right"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        ▷
      </button>
    </div>
  </div>
</div>
`;
